File: src/go/dp/countBits.go
package dp

func countBits(num int) []int {
	ans, s, n := make([]int, num+1), 0, 1
	ans[0] = 0
	for idx := 1; idx <= num; idx++ {
		if idx == n {
			ans[idx] = 1
			s = n
			n *= 2
			continue
		}
		ans[idx] = ans[s] + ans[idx-s]
	}
	return ans
}

File: algo-java/src/专项突破/整数/CountBits.java
package 专项突破.整数;

/**
 * @author Zexho
 * @date 2021/10/27 4:20 下午
 */
public class CountBits {
    public int[] countBits(int n) {
        int[] ans = new int[n + 1];
        int mark = 1, idx = 1;
        while (idx <= n) {
            ans[idx] = ans[idx - mark] + 1;
            idx++;
            if (idx == mark * 2) {
                mark = idx;
            }
        }
        return ans;
    }

}
